binder/src/androidTest/java/io/grpc/binder/internal/BinderClientTransportTest.java

```diff
@@ -334,41 +321,71 @@ public void transportInUse(boolean inUse) {
 
   private static final class TestStreamListener implements ClientStreamListener {
 
-    public StreamListener.MessageProducer messageProducer;
     public boolean ready;
     public Metadata headers;
-    @Nullable public Status closedStatus;
+
+    @GuardedBy("this")
+    private final Deque<MessageProducer> messageProducers = new ArrayDeque<>();
+
+    @GuardedBy("this")
+    @Nullable
+    private Status closedStatus;
 
     @Override
-    public void messagesAvailable(StreamListener.MessageProducer messageProducer) {
-      this.messageProducer = messageProducer;
+    public synchronized void messagesAvailable(StreamListener.MessageProducer messageProducer) {
+      messageProducers.add(messageProducer);
+      notifyAll();
     }
 
-    public synchronized void awaitMessages() {
-      while (messageProducer == null) {
-        try {
-          wait(100);
-        } catch (InterruptedException inte) {
-          throw new AssertionError("Interrupted waiting for messages");
+    /** Blocks until at least one MessageProducer has been provided for reading. */
+    public synchronized void awaitMessages() throws InterruptedException {
+      while (messageProducers.isEmpty()) {
+        wait();
+      }
+    }
+
+    /** Blocks until {@code n} messages can be produced (and discarded). */
+    public synchronized void readAndDiscardMessages(int n)
+        throws InterruptedException, IOException {
+      while (n > 0) {
+        while (closedStatus == null && messageProducers.isEmpty()) {
+          wait();
+        }
+        if (closedStatus != null) {
+          throw closedStatus.withDescription("premature close").asRuntimeException();
+        }
+        try (InputStream message = messageProducers.peek().next()) {
+          if (message == null) {
+            messageProducers.remove();
+            break;
```

Comment:
fixed thank you.

The tests passed despite this because all messages happen to arrive on a single MessageProducer today.

Reply:
You should add tests which would have caught this both in `readAndDiscardMessages` and `drainMessages`.  

The change also needs to be applied to the matching loop on line 385.

Reply:
My mistakes here were unfortunate and I appreciate you catching them. But it would be unusual to write unit tests for our unit tests as you request. The system-under-test here is `BinderClientTransport`, not `TestStreamListener`. This would be consistent with practice seen in other grpc-java tests, for example AbstractTransportTest.java defines a `ClientStreamListenerBase` that supports the various transport unit tests but doesn't have unit tests itself.

If you want, I can address your comment by filing a bug to factor out a universal FakeClientStreamListener to be used across grpc-java tests. As a 1st class fake, it *would* have unit tests of its own, and could replace these one-off listeners that today each unit test needs to write for itself.

LMK what you think.

Reply:
Good point.  Since it was only one file I didn't registered that it was embedded in a test class and you are right that tests for tests is ridiculous idea.
However, having a shared FakeClientStreamListener sounds like a great idea.  Thank you very much for offering.  I believe that the natural place to put it would be core/src/testFixtures where we already have a NoopClientStreamListener.

Reply:
Thanks! Filed https://github.com/grpc/grpc-java/issues/10845